Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,266 @@
+* Stack walking over the stack frames of the tail calls requires runtime assistance.
+ 
+# The new approach to tail calls using helpers
+## Objectives
+The new way of handling tail calls using helpers was designed with the following objectives:
+* It should be cheap to port to new platforms, architectures and code generators
+* It needs to work in both jitted and AOT compiled scenarios
+* It should support platforms where runtime code generation is not possible
+* The tail calls should be reasonably fast compared to regular calls with the same arguments
+* No runtime assistance should be necessary for unwinding stack with tail call frames on it
+* The stack should be unwindable at any spot during the tail calls to properly support sampling profilers and similar tools.
+* Stack walk during GC must be able to always correctly report GC references. 
+* It should work in all cases except those where a tail call is not allowed as described in the ECMA 335 standard section III.2.4
+
+## Requirements
+* The code generator needs to be able to compile a tail call to a target as a call to a thunk with the same parameters as the target, but void return, followed by a jump to an assembler helper.
+
+## Implementation
+This section describes the helper functions and data structures that the tail calls use and also describes the tail call sequence step by step. 
+### Helper functions
+The tail calls use the following thunks and helpers:
+* StoreArguments - this thunk has the same signature as the target function, except of the return type, which is void. It stores the arguments into a thread local storage together with the address of the corresponding CallTarget thunk and a descriptor of locations and types of managed references in the stored arguments data. This thunk is generated as IL and compiled by the jitter or AOT compiler. There is one such thunk per tail call target.
+* CallTarget - this thunk gets the arguments buffer that was filled by the StoreArguments thunk, loads the arguments from the buffer, releases the buffer and calls the target function. It is also generated as IL and compiled by the jitter or AOT compiler. There is one such thunk per tail call target.
+* TailCallHelper - this is an assembler helper that is responsible for restoring stack pointer to the location where it was when the first function in a tail call chain was entered and then jumping to the CallTarget thunk. This helper is common for all tail call targets.
+
+### Helper data structures
+The tail calls use the following data structures:
+* Thread local storage for arguments. It stores the arguments of a tail call for a short period of time between the StoreArguments and CallTarget calls.
+* Arguments GC descriptor - descriptor of locations and types of managed references in the arguments.
```

Comment:
> Arguments GC descriptor - descriptor of locations and types of managed references in the arguments. [](start = 1, length = 100)

On amd64 Unix, structs are passed directly in registers or on stack. Say the callee method tail called has struct args and those structs has GC ref/byref type fields (e.g. RuntimeTypeHandle).  Would GC descriptor contains details of gc ref/byref fields of the struct or  the GC-layout of value type needs to be constructed each time using its type handle during gc-reporting?  I am assuming the former.  Can you give some examples of what GC descriptor looks like while passing structs containing obj ref/byref fields?